### PR TITLE
Moved RBAC CRDs to group rbac.istio.io from config.istio.io

### PR DESCRIPTION
**Pull request by ymesika**
Refactor the CRD group names of the RBAC related resources from `config.istio.io/v1alpha2` to `rbac.istio.io/v1alpha1`.

Replaces #6820 

cc @rshriram

### CODE REVIEW

**Comment by ozevren** [comment]
Is there a migration story for deployments with existing resources?

**Comment by ymesika** [comment]
@rshriram that's a good point

**Comment by ozevren** [comment]
An approach like this can help support old and new API Groups side-by-side:
ozevren/istio@9ce843e

This is not tested, and still not complete enough to help with the RBAC case. For that, I think the right model would be to pass the full APIGroup/Version/Kind information instead of simply kinds.

File: install/kubernetes/helm/istio/charts/galley/templates/validatingwebhook.yaml [hunk 1]
```diff
@@ -33,6 +33,14 @@ webhooks:
         - httpapispecbindings
         - quotaspecs
         - quotaspecbindings
+      - operations:
```
Comment (ayj) [inline install/kubernetes/helm/istio/charts/galley/templates/validatingwebhook.yaml]: We can validate all kinds/versions once RBAC is in its own group (see below). The version/kind specific qualification was only required for `config.istio.io` which is split across mixer and pilot validation.

```yaml
- operations:
  - CREATE
  - UPDATE
  apiGroups:
  - rbac.istio.io
  apiVersions:
  - "*"
  resources:
  - "*"
```

File: install/kubernetes/helm/istio/charts/mixer/templates/clusterrole.yaml [hunk 2]
```diff
@@ -11,6 +11,9 @@ rules:
 - apiGroups: ["config.istio.io"] # istio CRD watcher
   resources: ["*"]
   verbs: ["create", "get", "list", "watch", "patch"]
+- apiGroups: ["rbac.istio.io"] # istio RBAC watcher
+  resources: ["*"]
+  verbs: ["create", "get", "list", "watch", "patch"]
```
Comment (ayj) [inline install/kubernetes/helm/istio/charts/mixer/templates/clusterrole.yaml]: Why does mixer need to `create` and `patch` RBAC resources?

File: install/kubernetes/helm/istio/charts/pilot/templates/clusterrole.yaml [hunk 3]
```diff
@@ -11,6 +11,9 @@ rules:
 - apiGroups: ["config.istio.io"]
   resources: ["*"]
   verbs: ["*"]
+- apiGroups: ["rbac.istio.io"]
+  resources: ["*"]
+  verbs: ["*"]
```
Comment (ayj) [inline install/kubernetes/helm/istio/charts/pilot/templates/clusterrole.yaml]: Pilot should need read-only access (i.e. `get`, `watch`, and `list` verbs)

**Comment by ozevren** [comment]
/hold

We should not check this in until the migration/compatibility story is addressed.

**Comment by ymesika** [comment]
Are there any other resources that needs to be migrated when upgrading from 0.8 to 1.0? Is there a general story for an upgrade or at the moment everything in 1.0 is compatible with 0.8?

**Comment by ayj** [comment]
Replacing `config.istio.io` with `rbac.istio.io` breaks seamless migration from 0.8 to 1.0. There will be a window during upgrade when RBAC policies will not be applied. Here are a few options:

1) ignore it. We accept that RBAC policies will be temporarily ineffective during upgrade

2) migrate RBAC policies from `config.istio.io` to `rbac.istio.io` _before_ upgrading to 1.0. This entails creating the new `rbac.istio.io` CRD, copying resources from `config.istio.io` to `rbac.istio.io`, installing the new Pilot/Mixer components, and finally removing the old `config.istio.io` RBAC CRD.

3) make Pilot/Mixer support both `config.istio.io` and `rbac.istio.io` for one release. User's have one release cycle to switch to the new group. We remove the old group in the next release.

#1 is a non-starter if people are using RBAC is production IMO. 

#2 gets the job done in one release cycle but requires more complex installer. #3 essentially does #2 but over two release cycles and doesn't require the installer.

Regardless of #2 or #3, we need to update references/samples/examples, mention it in the release notes, and consider creating a conversion command (analogous to [convert-networking-config](https://istio.io/docs/reference/commands/istioctl/#istioctl%20experimental%20convert-networking-config)).

**Comment by ymesika** [comment]
@ayj Thanks.
@rshriram Are you OK with #3? If you are I will make the required changes tomorrow and add another test to cover both group names. I don't see any issue supporting both CRDs beside logical conflicts but we can document it.

File: pilot/pkg/model/config.go [hunk 4]
```diff
@@ -533,7 +533,7 @@ var (
 		ClusterScoped: false,
 		Type:          "service-role-binding",
 		Plural:        "service-role-bindings",
-		Group:         "config",
+		Group:         "rbac",
```
Comment (yangminzhu) [inline pilot/pkg/model/config.go]: Could you also change the group name of RbacConfig in line 547? It's used to configure the global RBAC policy.

**Comment by yangminzhu** [comment]
Thanks for making this change, I'm a little confused about why do we need this rename? Is there any tech reason or is this just a naming convention for 1.0?

cc @liminw

**Comment by geeknoid** [comment]
/hold

Do we really need this change at this late stage in the release? Can we just live with the old name for 1.0 and fix for 1.1?

**Comment by rshriram** [comment]
RBAC is  unused so far in production. So I am not worried about migration, if and only if RBAC is the only thing sitting under config.istio.io. 

The name is extremely confusing. Given that we already have 20+ api objects, it’s very important to have a name that distinguishes the system over a generic “config.istio.io”. 

It’s also better to do this now, when no one is using it in production than postponing this by 3 months when there is a higher probability of someone using this in production and demanding backward compatibility.

File: pilot/pkg/model/config.go (hunk 4 above)
Comment (liminw) [reply pilot/pkg/model/config.go]: +1

**Comment by ayj** [comment]
[Option #1](https://github.com/istio/istio/pull/6874#issuecomment-403576482) is viable if RBAC is still alpha and no one is using in production.

**Comment by rshriram** [comment]
Lets go with Option 1. Better now than incurring backward compatibility/migration overhead in 1.1, and delaying release further.

**Comment by mandarjog** [comment]
I think this change makes sense, and #1 over others because it’s simpler.

That said there was an issue with cluster role / cluster role bindings being immutable, which causes upgrade headaches for k8s resources. Webhook clusterroles are updated here. If you can verify that this works, I am ok with the change.

**Comment by ymesika** [comment]
@mandarjog Welcome back. Anything else to test beside the RBAC tests in E2E?

**Comment by yangminzhu** [comment]
@ymesika This PR is to be merged to `istio:release-1.0`, how do we plan to merge it back to `master`?

**Comment by ymesika** [comment]
@yangminzhu I guess a merge from `release-1.0` to `master` will be done after the branch is closed. That's how it was done in `release-0.8`. There are quite a few PRs that got merged only to this release branch.
Are you trying to say there is some critical conflict with master? I initially requested a PR to `master` and there was no issue.

**Comment by yangminzhu** [comment]
@ymesika I'm concerned that this will create a conflict diverge between `1.0` and `master`, I'm going to add some e2e tests and additional sample configs (like #6749) to `master`.

**Comment by mandarjog** [comment]
Thanks @ymesika . did you test upgrade from 0.8 to 1.0 with this new change?

**Comment by yangminzhu** [comment]
@ymesika Would you mind to also update the config in `samples/bookinfo/kube/rbac`? I added them in #6749. Thank you!

**Comment by rshriram** [comment]
> Thanks @ymesika . did you test upgrade from 0.8 to 1.0 with this new change?

Upgrade with what? old RBAC config? or with istio 0.8 with no RBAC to istio 1.0 with no RBAC, just to see if there is some CRD issue?

**Comment by mandarjog** [comment]
I am not worried about upgrading istio rbac, like you others have said.

This is about changing k8s rbac resources, where is had seen “resource is immutable” errors in a prior version of k8s. I want to ensure that this change works ok with the current version of k8s.

**Comment by ymesika** [comment]
@mandarjog I see your concern with the upgrade but I don't think it's related to this PR.
When I install 0.8.0 and try to upgrade to current `release-1.0` (__not__ this PR branch) I get an error:
```sh
helm upgrade istio install/kubernetes/helm/istio --namespace istio-system

Error: UPGRADE FAILED: ClusterRoleBinding.rbac.authorization.k8s.io "istio-ingress-istio-system" is invalid: roleRef: Invalid value: rbac.RoleRef{APIGroup:"rbac.authorization.k8s.io", Kind:"ClusterRole", Name:"istio-ingress-istio-system"}: cannot change roleRef
```

Is this what you meant?

**Comment by liminw** [comment]
The error is unrelated. The error is for k8s RBAC. This PR is changing Istio RBAC api group.

**Comment by rshriram** [comment]
Fix lint

**Comment by ymesika** [comment]
@rshriram Not my fault

**Comment by ymesika** [comment]
@mandarjog After resolving other Helm upgrade issues (#7062) this PR upgrades from 0.8.0 to this branch without any error (using Helm).

**Comment by rshriram** [comment]
@liminw / @yangminzhu

**Comment by liminw** [comment]
Thanks for the change!

RBAC API is currently in v1alpha1 (https://github.com/istio/api/tree/master/rbac/v1alpha1). Should we change the version here to rbac.istio.io/v1alpha1?

**Comment by ymesika** [comment]
@liminw I noticed that but I didn't want to confuse users with a "downgrade" + it's only minor version.

**Comment by rshriram** [comment]
I think we should stick with the declared version in docs. its already a change for people to move from config.istio.io to rbac.istio.io.. so might as well do this

**Comment by ymesika** [comment]
@rshriram @liminw Let's nail it - The [docs URL](https://preliminary.istio.io/docs/reference/config/authorization/istio.rbac.v1alpha1/) (see URL path) are `v1alpha1` while within it the RBAC resources have API version `config.istio.io/v1alpha2`.
Should I keep it as it is now - `rbac.istio.io/v1alpha2` or update?

**Comment by yangminzhu** [comment]
The [docs URL](https://preliminary.istio.io/docs/reference/config/authorization/istio.rbac.v1alpha1/) is `v1alpha1` mainly because the RBAC protos are defined in the [v1alpha1 directory](https://github.com/istio/api/blob/2ac2e2b2c0f059b1eb2cdea367a1d48f0607ca42/rbac/v1alpha1/rbac.proto#L19). This is more a implementation detail while the resource API version `config.istio.io/v1alpha2` is a user facing definition.

We should move the protos to under a `v1alpha2` directory but this should be done after 1.0. It's too close to the release and not worth to do it right now.

Also we need to make sure the docs are updated accordingly once this PR is merged: the [concept page](https://preliminary.istio.io/docs/concepts/security/#authorization), the [task page](https://preliminary.istio.io/docs/tasks/security/role-based-access-control/) and the [reference page](https://preliminary.istio.io/docs/reference/config/authorization/istio.rbac.v1alpha1/) are needs to update.

Last, @ymesika could you help to test the `Authorization` task in the [spreadsheet](https://docs.google.com/spreadsheets/d/1mLf01qeJzrtDUhT_tz7uG6pqx2SpOUHRaOXz9ytILW4/edit#gid=1817277800) after this PR is merged? Thank you! (I'll also do a test)

**Comment by liminw** [comment]
I think our guideline is the config version should be consistent with our API version? @rshriram @geeknoid 

If that's the case, we should use v1alpha1. We already moved from config.istio.io to rbac.istio.io. Let's use the correct API version as well.

**Comment by rshriram** [comment]
I agree. since we are moving, we might as well fix both..

File: pilot/pkg/model/config.go [hunk 5]
```diff
@@ -463,7 +463,7 @@ var (
 	ServiceRole = ProtoSchema{
 		Type:        "service-role",
 		Plural:      "service-roles",
-		Group:       "config",
+		Group:       "rbac",
 		Version:     istioAPIVersion,
```
Comment (yangminzhu) [inline pilot/pkg/model/config.go]: The `istioAPIVersion` is actually `v1alpha2`, you have to change the `ServiceRole`, `ServiceRoleBinding` and `RbacConfig` to use a separate version string if you want to change it to `v1alpha1`.

Comment (ymesika) [reply pilot/pkg/model/config.go]: Fixed, Thanks.

Comment (ymesika) [reply pilot/pkg/model/config.go]: BTW, I merged you e2e test updates from revision-1.0 to make sure they are passing too.

**Comment by rshriram** [comment]
@yangminzhu / @liminw is this okay to merge? if so, please approve the PR

**Pull request merged**